### PR TITLE
 Introducing an additional tiles scale factor and possible fix for #776

### PR DESCRIPTION
**Pull request by secaw**
Based on what @roman-nevr proposed in #728, it is now possible to set an additional tiles scale factor for map scaling. This now works both for ScaledToDpi and standard size display.

Also removed redundant `MapTileProviderBase.clearTileCache()` from `MapTileProviderArray.detach()` and `MapView.detach()` and moved it to `MapTileProviderBase.detach()` where it should belong. Might be a fix for #776.

### CODE REVIEW

**Comment by monsieurtanuki** [comment]
@spyhunter99 Wouldn't that make sense to include it into 6.0.0 - I mean more specifically the cache part? Or is it too late?

**Comment by miguel-porto** [comment]
Note that I'm facing frequent silent freezes when MapView.onDetach() is called. Maybe it's related to this.

**Comment by monsieurtanuki** [comment]
@miguel-porto What you're experiencing is specifically with these commits, or mine in #908?
I just created #919 in order to improve memory cache management - and incidentally fix issues. And I don't think my own #908 goes 100% in the right direction.

**Comment by miguel-porto** [comment]
I'm experiencing the freezes both with the master branch and with your PR #908. I didn't try this one here. Sorry I haven't been able to keep up so frequently in testing, as I have another project that is priority. Your #919, at first glance, seems to be the cure to my problems!

**Comment by secaw** [comment]
@miguel-porto Would be absolutely great, if you find the time to test this PR!

@spyhunter99 Like @monsieurtanuki I would recommend to include at least the cache part of this PR in 6.0.0 (if it's still possible) as it's cleaning up the detach chain regarding `MapTileProviderBase.clearTileCache()` in a more logical and safer way. Note that I have this released for a few days now and #776 hasn't reoccured yet as well as some other ANRs seem to have disappeared which cause I wasn't able to identify. So I have good reason to believe that this actually fixes them!

**Comment by spyhunter99** [comment]
If you all are reasonably confident that this solves the problem, then that's fine. I've never seen the slight pause on this end. If it can be replicated, the best thing to do is to use the android device monitor and run a profiling session to find out what methods are causing the pausing. It's really the only way

**Comment by spyhunter99** [comment]
I'm on @secaw 's branch and i'm now seeing cases where by the entire screen flips grey the almost immediately back to tiles and i'm seeing pauses when crossing integer. But...after a short amount of time of pinch zooming and whatnot, whatever the problem was seemed to go away. I occasionally still see a brief instant where the screen is completely gray. Profiling the session gives the following

image

Just looking at the output, a surprising amount of time is spent loading tiles from the database. This could indicate that the in memory cache is too small or that we may have too many threads running that are utilizing the database. Note that the above was running on the genymotion vm.

Profiling by my s5, its quite a bit different. I never saw the flicker, but i did see the pausing while zooming.
image
although there's a lot more tile spent downloading tiles, which is normal. Again, most of the time was spent loading a tile from the sql cache.

Reducing the threads to 4 didn't seem to help the problem at all. It's possible we are just seeing the garbage collector running

**Comment by spyhunter99** [comment]
I'm also seeing the hang on detach issue.  Ugh, gonna check master for the hang later

**Comment by secaw** [comment]
@spyhunter99 That's strange, thanks for testing. I don't think, what you're seeing is caused by this PR though. Did you back check the master again?
This here is not about the slight pause thing and shouldn't have an influence on tile display / loading at all. I cannot reproduce any of that nor the hang on detach issue.
For me it seems to fix #776 though and possibly other related ANRs caused by concurrency issues before, that I believe now had the same source (basically `MapTileProviderBase.TileCache` was cleared too early in the detach chain while it was still open to be written to).
Just wondering, were you building your own `MapTileProviderArray` for the test above?

**Comment by miguel-porto** [comment]
@secaw so far so good. This code here doesn't seem to hang on detach any more. But I'll keep on trying, since the hang problem was unpredictable, so it's better to keep on testing a bit more. I'll post here after I am more certain that this solves the issue.

**Comment by miguel-porto** [comment]
Oh sorry speaked too early... yes the hang-on-detach problem still happens. I'm looking forward to #920...

**Comment by miguel-porto** [comment]
Note that the hang-on-detach is silent, no exceptions are thrown, it gets completely blocked and the activity won't close. So it may be a different issue from #776.

**Comment by spyhunter99** [comment]
Do we have a solid reproducible setup for the hang on detach issue? Is it only with certain overlays? when tiles are still in the download queue? maybe during a zoom animation?

**Comment by miguel-porto** [comment]
I still find it unpredictable, but it seems to happen only after I have zoomed in & out a lot, so, after the memory cache being used a lot. It happens offline (not sure if online also). Note that I have nearly 1 GB of cached tiles. I'm only using Bing maps and SimpleFastPointOverlay (but as far as I remember, it also happens without any overlay).
It only happens when MapView.onDetach() is called. It never happened during the app use, even if I zoom like crazy. I'll post here if I have more clues. But can you reproduce it?

**Comment by spyhunter99** [comment]
@miguel-porto i'm looking at the code for detach now... here's the basic order of operations

 - kill all overlays
 - destroy the tile provider (suspect)
 - purge the tile provider's memory cache (suspect)
 - destroy the tile request handler
 - destroy the projection

i'd suggest putting log statements in between each line to try and isolate the root cause.

**Comment by spyhunter99** [comment]
do we have consensus on this yet? I'm ok with merging this

**Comment by secaw** [comment]
@spyhunter99 Still no #776 or related ANRs since this PR included in published app, so I'm even more positive about this!

@miguel-porto Thanks for testing, too bad it doesn't solve your hang on detach also. Might be good to open a separate issue.
I can't reproduce the hang on detach at all, although I'm switching back and forth the map activity a lot. Maybe we should compare setups, e.g. I'm just using Itemized Overlays, Paths and an additional map overlay. No compass, minimap etc. Did you check deactivating such additional features, if present in your setup?

**Comment by miguel-porto** [comment]
It seems that the hang-on-detach is fixed by #908 ! Still not fully sure, but I'm positive about it.
My setup is pretty simple, just Bing maps and SimpleFastPointOverlay on top of it. However, I have 1GB of cached tiles and work offline, it might have something to do with pushing memory to the limits. But no worries with this for now, because the last commits to #908 probably fixed it!
